contrib/nydus-backend-proxy/src/main.rs

```diff
@@ -1,250 +1,275 @@
+// Copyright (C) 2022 Alibaba Cloud. All rights reserved.
+//
+// SPDX-License-Identifier: (Apache-2.0 AND BSD-3-Clause)
+
 #[macro_use]
 extern crate rocket;
-extern crate http_range;
-extern crate nix;
+#[macro_use]
+extern crate lazy_static;
 #[macro_use(crate_authors, crate_version)]
 extern crate clap;
 
-use clap::{App, Arg};
-use rocket::fs::FileServer;
+use std::collections::HashMap;
 use std::env;
-use std::path::PathBuf;
+use std::os::unix::io::AsRawFd;
+use std::os::unix::prelude::RawFd;
+use std::path::{Path, PathBuf};
+use std::{fs, io};
 
-mod blob_backend {
-    use http_range::HttpRange;
-    use nix::sys::uio;
-    use once_cell::sync::OnceCell;
-    use rocket::fs::NamedFile;
-    use rocket::http::Status;
-    use rocket::request::{self, FromRequest, Outcome};
-    use rocket::response::{self, stream::ReaderStream, Responder};
-    use rocket::Request;
-    use rocket::Response;
-    use std::collections::HashMap;
-    use std::os::unix::io::AsRawFd;
-    use std::path::{Path, PathBuf};
-    use std::{ffi, fs, io};
+use clap::{App, Arg};
+use nix::sys::uio;
 
-    #[derive(Debug)]
-    struct BlobBackend {
-        root: PathBuf,
-        blobs: HashMap<ffi::OsString, fs::File>,
-    }
-    impl BlobBackend {
-        fn new(root: PathBuf) -> Self {
-            Self {
-                blobs: get_blobmap(&root),
-                root,
-            }
-        }
-    }
-    fn get_blobmap(path: &Path) -> HashMap<ffi::OsString, fs::File> {
-        let mut blobfdmap = HashMap::new();
-        for entry in path.read_dir().expect("read blobsdir failed").flatten() {
+use http_range::HttpRange;
+
+use rocket::fs::{FileServer, NamedFile};
+use rocket::futures::lock::{Mutex, MutexGuard};
+use rocket::http::Status;
+use rocket::request::{self, FromRequest, Outcome};
+use rocket::response::{self, stream::ReaderStream, Responder};
+use rocket::{Request, Response};
+
+lazy_static! {
+    static ref BLOB_BACKEND: Mutex<BlobBackend> = Mutex::new(BlobBackend {
+        root: PathBuf::default(),
+        blobs: HashMap::new()
+    });
+}
+
+async fn blob_backend_mut() -> MutexGuard<'static, BlobBackend> {
+    BLOB_BACKEND.lock().await
+}
+
+async fn init_blob_backend(root: &Path) {
+    let mut b = BlobBackend {
+        root: root.to_path_buf(),
+        blobs: HashMap::new(),
+    };
+
+    b.populate_blobs_map();
+    *BLOB_BACKEND.lock().await = b;
+}
+
+#[derive(Debug)]
+struct BlobBackend {
+    root: PathBuf,
+    blobs: HashMap<String, fs::File>,
+}
+impl BlobBackend {
+    fn populate_blobs_map(&mut self) {
```

Comment:
Is it necessary to load the whole blob directory if the blob is not found on fetching?

Reply:
My thought is to load all new files into fd map after re-scanning since the `nydusd-backend-proxy` 's  exports all the files  int `blobsdir` , which I think is its SLO